### PR TITLE
Better error messages for Type Errors

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
This PR intends to improve the error messages in case of type errors and also introduces stricter type checking for primitive types.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
fixes https://github.com/lyft/flyte/issues/1439
fixes https://github.com/lyft/flyte/issues/1423

## Follow-up issue
_NA_

### CODE REVIEW

**Review by wild-endeavor — COMMENTED** [review]
can we test one of these errors on console?  just want to see what it looks like

File: flytekit/core/base_task.py [hunk 1]
```diff
@@ -501,12 +503,12 @@ def dispatch_execute(
...
                 if isinstance(v, tuple):
-                    raise AssertionError(f"Output({k}) in task{self.name} received a tuple {v}, instead of {py_type}")
+                    raise TypeError(f"Output({k}) in task{self.name} received a tuple {v}, instead of {py_type}")
                 try:
                     literals[k] = TypeEngine.to_literal(exec_ctx, v, py_type, literal_type)
                 except Exception as e:
-                    logger.error(f"failed to convert return value for var {k} with error {type(e)}: {e}")
-                    raise e
+                    logger.error(f"Failed to convert return value for var {k} with error {type(e)}: {e}")
+                    raise TypeError(f"Failed to convert return value for var {k} for function {self.name}") from e
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/core/base_task.py:511]:
```
Traceback (most recent call last):
  File "/Users/kevin/git/flytekit/flytekit/core/base_task.py", line 508, in dispatch_execute
    literals[k] = TypeEngine.to_literal(exec_ctx, v, py_type, literal_type)
  File "/Users/kevin/git/flytekit/flytekit/core/type_engine.py", line 390, in to_literal
    transformer.assert_type(python_type, python_val)
  File "/Users/kevin/git/flytekit/flytekit/core/type_engine.py", line 65, in assert_type
    raise TypeError(f"Type of Val '{v}' is not an instance of {t}")
TypeError: Type of Val 'Welcome, kevin!' is not an instance of <class 'int'>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 28, in <module>
    print("Final result", welcome(name="kevin"))
  File "/Users/kevin/git/flytekit/flytekit/core/workflow.py", line 236, in __call__
    return flyte_entity_call_handler(self, *args, **input_kwargs)
  File "/Users/kevin/git/flytekit/flytekit/core/promise.py", line 876, in flyte_entity_call_handler
    result = cast(LocallyExecutable, entity).local_execute(child_ctx, **kwargs)
  File "/Users/kevin/git/flytekit/flytekit/core/workflow.py", line 251, in local_execute
    function_outputs = self.execute(**kwargs)
  File "/Users/kevin/git/flytekit/flytekit/core/workflow.py", line 686, in execute
    return exception_scopes.user_entry_point(self._workflow_function)(**kwargs)
  File "/Users/kevin/git/flytekit/flytekit/common/exceptions/scopes.py", line 198, in user_entry_point
    return wrapped(*args, **kwargs)
  File "test.py", line 22, in welcome
    greeting = greet(name=name)
  File "/Users/kevin/git/flytekit/flytekit/core/base_task.py", line 274, in __call__
    return flyte_entity_call_handler(self, *args, **kwargs)
  File "/Users/kevin/git/flytekit/flytekit/core/promise.py", line 869, in flyte_entity_call_handler
    return cast(LocallyExecutable, entity).local_execute(ctx, **kwargs)
  File "/Users/kevin/git/flytekit/flytekit/core/base_task.py", line 255, in local_execute
    outputs_literal_map = self.dispatch_execute(ctx, input_literal_map)
  File "/Users/kevin/git/flytekit/flytekit/core/base_task.py", line 511, in dispatch_execute
    raise TypeError(f"Failed to convert return value for var {k} for function {self.name}") from e
TypeError: Failed to convert return value for var o0 for function __main__.greet
```

[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/core/base_task.py:511]: An example is here.
https://gist.github.com/pingsutw/0b45f5413dcc11fb8c6a7c6b640dd5ad

**Review by pingsutw — APPROVED** [review]

File: tests/flytekit/unit/core/test_type_hints.py [hunk 2]
```diff
@@ -1338,3 +1340,16 @@ def t2() -> dict:
...
+def test_error_messages():
+    @task
+    def foo(a: int, b: str) -> typing.Tuple[int, str]:
+        return 10, "hello"
+
+    @task
+    def foo2(a: int, b: str) -> typing.Tuple[int, str]:
+        return "hello", 10
+
+    with pytest.raises(TypeError):
```
Comment (eapolinario) [inline tests/flytekit/unit/core/test_type_hints.py:1354]: can you also use the `match` parameter for this test? Just to confirm that the message is what we expect.

File: tests/flytekit/unit/core/test_type_hints.py [hunk 3]
```diff
@@ -1338,3 +1340,16 @@ def t2() -> dict:
     expected_struct = Struct()
     expected_struct.update({"k1": "v1", "k2": 3, "4": {"one": [1, "two", [3]]}})
     assert output_lm.literals["o0"].scalar.generic == expected_struct
+
+
+def test_error_messages():
```
Comment (eapolinario) [inline tests/flytekit/unit/core/test_type_hints.py:1345]: can we have a test where the mismatch happens in a proto? We had an issue in the [OSS slack ](https://flyte-org.slack.com/archives/CREL4QVAQ/p1631818531170100?thread_ts=1630642686.070000&cid=CREL4QVAQ) where a workflow input of type Dictionary but it was being passed a list of dictionaries (and as you can see from the slack message, the error is not intuitive).

**Review by eapolinario — COMMENTED** [review]
This is major step up in usability. Thanks a lot for working on it!

**Review by eapolinario — APPROVED** [review]
